tests/integration/test_runner_manager_openstack.py

```diff
@@ -0,0 +1,411 @@
+
+    cloud_config = OpenStackCloudConfig(
+        clouds_config=clouds_config,
+        cloud="testcloud",
+    )
+    server_config = OpenStackServerConfig(
+        image=openstack_test_image,
+        flavor=flavor_name,
+        network=network_name,
+    )
+    runner_config = GitHubRunnerConfig(
+        github_path=github_path,
+        labels=["openstack_test", runner_label],
+    )
+    service_config = SupportServiceConfig(
+        proxy_config=proxy_config,
+        dockerhub_mirror=None,
+        ssh_debug_connections=None,
+        repo_policy_url=None,
+        repo_policy_token=None,
+    )
+    return OpenstackRunnerManager(
+        app_name, cloud_config, server_config, runner_config, service_config
+    )
+
+
+@pytest_asyncio.fixture(scope="module", name="runner_manager")
+async def runner_manager_fixture(
+    openstack_runner_manager: OpenstackRunnerManager,
+    token: str,
+    github_path: GithubPath,
+    log_dir_base_path: dict[str, Path],
+) -> RunnerManager:
+    """Get RunnerManager instance.
+
+    Import of log_dir_base_path to monkeypatch the runner logs path with tmp_path.
+    """
+    config = RunnerManagerConfig(token, github_path)
+    return RunnerManager(openstack_runner_manager, config)
+
+
+@pytest_asyncio.fixture(scope="function", name="runner_manager_with_one_runner")
+async def runner_manager_with_one_runner_fixture(runner_manager: RunnerManager) -> RunnerManager:
+    runner_manager.create_runners(1)
+    runner_list = runner_manager.get_runners()
+    try:
+        await assert_runner_amount(runner_manager, 1)
+    except TimeoutError as err:
+        raise AssertionError("Test arrange failed: Expect one runner") from err
+
+    runner = runner_list[0]
+    assert (
+        runner.cloud_state == CloudRunnerState.ACTIVE
+    ), "Test arrange failed: Expect runner in active state"
+    try:
+        await wait_for(
+            lambda: runner_manager.get_runners()[0].github_state == GitHubRunnerState.IDLE,
+            timeout=120,
+            check_interval=10,
+        )
+    except TimeoutError as err:
+        raise AssertionError("Test arrange failed: Expect runner in idle state") from err
+    return runner_manager
+
+
+def workflow_is_status(workflow: Workflow, status: str) -> bool:
+    """Check if workflow in provided status.
+
+    Args:
+        workflow: The workflow to check.
+        status: The status to check for.
+
+    Returns:
+        Whether the workflow is in the status.
+    """
+    workflow.update()
+    return workflow.status == status
+
+
+async def assert_runner_amount(runner_manager: RunnerManager, num: int):
+    """Assert the number of runner a runner manager has.
+
+    A TimeoutError will be thrown if runners are still found after timeout.
+
+    Args:
+        runner_manager: The RunnerManager to check.
+        num: Number of runner to check for.
+    """
+    runner_list = runner_manager.get_runners()
+    assert isinstance(runner_list, tuple)
```

Comment:
Just out of curiosity, may I know what the instance check does here?

Reply:
It is just type checking the return type is tuple. Since we have older function that returns list where tuple is a better fit.
This is just to ensure we would not still use tuple as the return type after future refactors.